### PR TITLE
Enable device change during training + minor forward() refactoring

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
During training loop, users may want to switch devices. This PR adds support for such scenario

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ def __init__(self, module):
         self._export_again = False
         # TODO: This is incorrect when different layers may be in different devices
         self._device = next(module.parameters()).device
-        self._require_export = False
+        self._device_changed = False
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule.py:100]: The only time export is required is when the model has never been exported before or if the device has changed. Is my understanding correct?
Do you think we could have other reasons in the future to require a complete export (maybe in that case we could keep the name the same as it was ```_require_export```?)

Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule.py:100]: We don't actually re-export the ONNX graph when the device changes. We just recreate the InferenceSession
This flag was poorly named

**Review by baijumeswani — DISMISSED** [review]
Looks good.

**Review by baijumeswani — APPROVED** [review]
Looks good.

**Pull request merged**